ramips: add support for Mikrotik LtAP

db4rne opened the pull request:
ramips: add support for Mikrotik LtAP-2HnD

Mikrotik LtAP-2HnD is a outdoor/car WLAN 4 router with integrated GPS receiver and two mPCIe slots.

Specifications:
* SoC: MT7621A
* RAM: 128 MiB Nanya NT5CC64M16GP-DI
* Flash: 16 MiB winbond W25Q128JV
* WLAN: 
  * Atheros AR9382 with power amplifier SKY 85330 (2x2 internal antennas, with RF switches for external connectors)
* Ethernet: 1 Gbps, single port
* USB Host: USB 2.0 Speeds
* Serial: 115200 baud
* LEDs: Power, System, GPS, 5* RSSI
*  mPCIe:
   * miniPCIe slot 1: PCIe and USB 2.0 Host (via switch shared with USB Host interface on outside)
   * miniPCIe slot 2: USB 2.0 and 3.0
* SIM Cards:
  * Slot 1 Connected to mPCIe slot 1
  * Slot 2 and 3 connected to mPCIe slot 2 via switch
* GPS: MTK 3333 on serial port 2 (/dev/ttyS1), 115200 baud and PPS on pin gpiochip0, pin 14

Installation:

1. TFTP boot openwrt
2. sysupgrade

Commit:
ramips: add support for Mikrotik LtAP-2HnD

Mikrotik LtAP-2HnD is a outdoor/automotive WLAN 4 router with integrated GPS
receiver and two mPCIe slots.

Specifications:
* SoC: MT7621A
* RAM: 128 MiB Nanya NT5CC64M16GP-DI
* Flash: 16 MiB winbond W25Q128JV
* WLAN:
  * Atheros AR9382 with power amplifier SKY 85330 (2x2 internal antennas,
    with RF switches for external connectors)
* Ethernet: 1 Gbps, single port
* USB Host: USB 2.0 Speeds
* Serial: 115200 baud
* LEDs: Power, System, GPS, 5* RSSI
*  mPCIe:
   * miniPCIe slot 1: PCIe and USB 2.0 Host (via switch shared with USB Host)
   * miniPCIe slot 2: USB 2.0 and 3.0
* SIM Cards:
  * Slot 1 Connected to mPCIe slot 1
  * Slot 2 and 3 connected to mPCIe slot 2 via switch
* GPS: MTK 3333 on serial port 2 (/dev/ttyS1), 115200 baud and PPS on gpio 14

gpios are exposed to /sys/class/gpio:

* usb-select: swithes USB 2.0 interface between external port and internal
mPCIe slot 1 default is the external USB interface
* gps-reset: resets the GPS interface chip
* sim-select: switches between sim slot 2 and 3 connected to mPCIe slot 2
* gps-ant-select: switches GPS antenna between internal antenna and SMA
connected antenna
* lte-reset: resets mPCIe slot 2

Flashing:
 TFTP boot initramfs image and then perform sysupgrade. Follow common
 MikroTik procedure as in https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Arne Zachlod <arne@nerdkeller.org>

db4rne commented:
TODO:

- [ ] clean up firmware script
- [ ] find GPIOS:
   - [ ] GPS LED
   - [ ] RSSI LEDs
   - [ ] WLAN antenna switches
   - [ ] SIM 2/3 switch
- [ ] check LNA if usage enabled
- [ ] check PA if usage enabled
- [ ] clean up DTS file

john-tho commented:
I've not yet looked through any of your changes. Thank you for working on support for this device. I saw that the [RouterOS v7 GPL dump](https://github.com/Vlad1mir-D/routeros-7-source) has some details that might help you with GPIOs.

It doesn't use normal LED drivers, but their `leds-rb` driver is included in their linux patch. Guessing the type *wifi* LEDs should use a different GPIO parent—the wifi chip, but who knows.
See the dropdowns below.

<details>
<summary>RouterOS mt7621.dts</summary>
</br>

```
diff -puNrb linux-5.6.3/arch/mips/boot/dts/rb/mt7621.dts linux/arch/mips/boot/dts/rb/mt7621.dts
--- linux-5.6.3/arch/mips/boot/dts/rb/mt7621.dts	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/mips/boot/dts/rb/mt7621.dts	2022-01-04 10:36:49.711252814 +0200
@@ -0,0 +1,542 @@
+#include <dt-bindings/interrupt-controller/mips-gic.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/dts-v1/;
+
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+	compatible = "mediatek,mtk7621-soc";
+
+	cpus {
+		cpu@0 {
+			compatible = "mips,mips1004Kc";
+		};
+
+		cpu@1 {
+			compatible = "mips,mips1004Kc";
+		};
+	};
+
+	cpuintc: cpuintc@0 {
+		#address-cells = <0>;
+		#interrupt-cells = <1>;
+		interrupt-controller;
+		compatible = "mti,cpu-interrupt-controller";
+	};
+
+	cpuclock: cpuclock@0 {
+		#clock-cells = <0>;
+		compatible = "fixed-clock";
+	};
+
+	hclk: hclk@0 {
+	        #clock-cells = <0>;
+		compatible = "fixed-clock";
+
+		clock-frequency = <220000000>;
+	};
+
+	sysclock: sysclock@0 {
+		#clock-cells = <0>;
+		compatible = "fixed-clock";
+
+		clock-frequency	= <50000000>;
+	};
+
+	rb760-leds {
+		status = "disabled";
+		compatible = "leds-rb";
+
+		sfp_led {
+			label = "sfp-led";
+			gpios = <&gpio0 9 GPIO_ACTIVE_HIGH>;
+		};
+		usb_power_off {
+			label = "usb-power-off";
+			gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
+		};
+		power_led {
+			label = "power-led";
+			gpios = <&gpio0 16 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+		button {
+			label = "button";
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+			default-state = "input";
+		};
+		mode_button {
+			label = "mode-button";
+			gpios = <&gpio0 13 GPIO_ACTIVE_LOW>;
+			default-state = "input";
+		};
+	};
+
+	m33-leds {
+		status = "disabled";
+		compatible = "leds-rb";
+
+		user_led {
+			label = "user-led";
+			gpios = <&gpio0 0 GPIO_ACTIVE_HIGH>;
+		};
+		ext_pin3 {
+			label = "ext-pin3";
+			gpios = <&gpio0 3 GPIO_ACTIVE_HIGH>;
+			default-state = "input";
+		};
+		ext_pin5 {
+			label = "ext-pin5";
+			gpios = <&gpio0 4 GPIO_ACTIVE_HIGH>;
+			default-state = "input";
+		};
+		pcie_power_off {
+			label = "mpcie-power-off";
+			gpios = <&gpio0 9 GPIO_ACTIVE_LOW>;
+		};
+		pcie2_power_off {
+			label = "mpcie2-power-off";
+			gpios = <&gpio0 10 GPIO_ACTIVE_LOW>;
+		};
+		usb_power_off {
+			label = "usb-power-off";
+			gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
+		};
+		button {
+			label = "button";
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+			default-state = "input";
+		};
+	};
+
+	m11-leds {
+		status = "disabled";
+		compatible = "leds-rb";
+
+		user_led {
+			label = "user-led";
+			gpios = <&gpio0 0 GPIO_ACTIVE_HIGH>;
+		};
+		pcie_power_off {
+			label = "mpcie-power-off";
+			gpios = <&gpio0 9 GPIO_ACTIVE_LOW>;
+		};
+		button {
+			label = "button";
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+			default-state = "input";
+		};
+		led1 {
+			label = "led1";
+			gpios = <&gpio0 26 GPIO_ACTIVE_LOW>;
+		};
+		led2 {
+			label = "led2";
+			gpios = <&gpio0 25 GPIO_ACTIVE_LOW>;
+		};
+		led3 {
+			label = "led3";
+			gpios = <&gpio0 24 GPIO_ACTIVE_LOW>;
+		};
+		led4 {
+			label = "led4";
+			gpios = <&gpio0 23 GPIO_ACTIVE_LOW>;
+		};
+		led5 {
+			label = "led5";
+			gpios = <&gpio0 22 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	ltap-leds {
+		status = "disabled";
+		compatible = "leds-rb";
+
+		user_led {
+			label = "user-led";
+			gpios = <&gpio0 0 GPIO_ACTIVE_HIGH>;
+		};
+		pcie_power_off {
+			label = "mpcie-power-off";
+			gpios = <&gpio0 10 GPIO_ACTIVE_LOW>;
+		};
+		pcie2_power_off {
+			label = "mpcie2-power-off";
+			gpios = <&gpio0 11 GPIO_ACTIVE_LOW>;
+		};
+		usb_power_off {
+			label = "usb-power-off";
+			gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
+		};
+		gps_reset {
+			label = "gps-reset";
+			gpios = <&gpio0 13 GPIO_ACTIVE_HIGH>;
+			default-state = "keep";
+		};
+		control {
+			label = "control";
+			gpios = <&gpio0 15 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+		button {
+			label = "button";
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+			default-state = "input";
+		};
+		power_led {
+			label = "power-led";
+			gpios = <&gpio0 23 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+		mode_button {
+			label = "mode-button";
+			gpios = <&gpio0 24 GPIO_ACTIVE_LOW>;
+			default-state = "input";
+		};
+		sim_select {
+			label = "sim-select";
+			gpios = <&gpio0 25 GPIO_ACTIVE_LOW>;
+		};
+		gps_ant_select {
+			label = "gps-ant-select";
+			gpios = <&gpio0 27 GPIO_ACTIVE_HIGH>;
+		};
+		lte_reset {
+			label = "lte-reset";
+			gpios = <&gpio0 29 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+		led1 {
+			label = "led1";
+			gpios = <&gpio0 0 GPIO_ACTIVE_HIGH>;
+			type = "wifi";
+		};
+		led2{
+			label = "led2";
+			gpios = <&gpio0 1 GPIO_ACTIVE_HIGH>;
+			type = "wifi";
+		};
+		led3 {
+			label = "led3";
+			gpios = <&gpio0 2 GPIO_ACTIVE_HIGH>;
+			type = "wifi";
+		};
+		led4 {
+			label = "led4";
+			gpios = <&gpio0 3 GPIO_ACTIVE_HIGH>;
+			type = "wifi";
+		};
+		led5 {
+			label = "led5";
+			gpios = <&gpio0 4 GPIO_ACTIVE_HIGH>;
+			type = "wifi";
+		};
+	};
+	
+	beeper {
+	        status = "disabled";
+
+	        compatible = "rb,gpio-beep";
+
+		gpios = <&gpio0 15 GPIO_ACTIVE_LOW>;
+	};
+
+	poe-simple {
+		status = "disabled";
+		compatible = "rb,poe-out-simple";
+		eth_port = <4>;
+		has_low_current_sense = <1>;
+		gpi_status = <&gpio0 7 GPIO_ACTIVE_HIGH>;
+		gpo_on = <&gpio0 17 GPIO_ACTIVE_HIGH>;
+	};
+
+	palmbus@1e000000 {
+		compatible = "palmbus";
+		reg = <0x1E000000 0x100000>;
+		ranges = <0x0 0x1E000000 0x0FFFFF>;
+
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		sysc@0 {
+			compatible = "mtk,mt7621-sysc";
+			reg = <0x0 0x100>;
+		};
+
+		wdt@100 {
+			compatible = "mtk,mt7621-wdt";
+			reg = <0x100 0x100>;
+		};
+
+		gpio0: gpio@600 {
+			#gpio-cells = <2>;
+			#interrupt-cells = <2>;
+			compatible = "mediatek,mt7621-gpio";
+			gpio-controller;
+			interrupt-controller;
+			reg = <0x600 0x100>;
+			interrupt-parent = <&gic>;
+			interrupts = <GIC_SHARED 12 IRQ_TYPE_LEVEL_HIGH>;
+		};
+
+		memc@5000 {
+			compatible = "mtk,mt7621-memc";
+			reg = <0x300 0x100>;
+		};
+
+		cpc@1fbf0000 {
+			     compatible = "mtk,mt7621-cpc";
+			     reg = <0x1fbf0000 0x8000>;
+		};
+
+		mc@1fbf8000 {
+			    compatible = "mtk,mt7621-mc";
+			    reg = <0x1fbf8000 0x8000>;
+ 		};
+
+		uartlite@c00 {
+			compatible = "ns16550a";
+			reg = <0xc00 0x100>;
+
+			clocks = <&sysclock>;
+
+			interrupt-parent = <&gic>;
+			interrupts = <GIC_SHARED 26 IRQ_TYPE_LEVEL_HIGH>;
+
+			reg-shift = <2>;
+			reg-io-width = <4>;
+			no-loopback-test;
+		};
+
+		uartlite@e00 {
+			status = "disabled";
+			compatible = "ns16550a";
+			reg = <0xe00 0x100>;
+
+			clocks = <&sysclock>;
+
+			interrupt-parent = <&gic>;
+			interrupts = <GIC_SHARED 28 IRQ_TYPE_LEVEL_HIGH>;
+
+			reg-shift = <2>;
+			reg-io-width = <4>;
+			no-loopback-test;
+		};
+		
+		spi@b00 {
+			status = "okay";
+
+			compatible = "ralink,mt7621-spi";
+			reg = <0xb00 0x100>;
+
+			clocks = <&hclk>;
+
+			resets = <&rstctrl 18>;
+			reset-names = "spi";
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			m25p80@0 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				status = "disabled";
+				compatible = "jedec,spi-nor";
+				reg = <0>;
+				spi-max-frequency = <30000000>;
+				m25p,chunked-io = <32>;
+	
+				partition@40000 {
+					 label = "RouterOS";
+					 reg = <0x40000 0xfc0000>;
+				};
+				partition@0 {
+					 label = "RouterBoot";
+					 reg = <0x0 0x40000>;
+				};
+			};
+
+			m25p80@1 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				status = "disabled";
+				compatible = "jedec,spi-nor";
+				reg = <1>;
+				spi-max-frequency = <30000000>;
+				m25p,chunked-io = <32>;
+	
+				partition@40000 {
+					 label = "RouterOS";
+					 reg = <0x40000 0xfc0000>;
+				};
+				partition@0 {
+					 label = "RouterBootFake";
+					 reg = <0x0 0x40000>;
+				};
+			};
+			
+			m25p80_boot@0 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				status = "disabled";
+				compatible = "jedec,spi-nor";
+				reg = <0>;
+				spi-max-frequency = <30000000>;
+				m25p,chunked-io = <32>;
+	
+				partition@0 {
+					 label = "RouterBoot";
+					 reg = <0x0 0x40000>;
+				};
+			};
+			
+			ts@1 {
+				status = "disabled";
+				
+				compatible = "rb-spi-ts";
+				reg = <1>;
+				spi-max-frequency = <2000000>;
+			};
+		};
+	};
+
+	rstctrl: rstctrl {
+		compatible = "ralink,rt2880-reset";
+		#reset-cells = <1>;
+	};
+
+	sdhci@1e130000 {
+		status = "disabled";
+
+		compatible = "ralink,mt7620-sdhci";
+		reg = <0x1E130000 0x4000>;
+
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_SHARED 20 IRQ_TYPE_LEVEL_HIGH>;
+	};
+
+	xhci@1e1c0000 {
+		status = "disabled";
+
+		compatible = "mediatek,mt8173-xhci";
+		reg = <0x1e1c0000 0x1000
+		       0x1e1d0700 0x0100>;
+		reg-names = "mac", "ippc";
+
+		clocks = <&sysclock>;
+		clock-names = "sys_ck";
+
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_SHARED 22 IRQ_TYPE_LEVEL_HIGH>;
+	};
+
+	crypto@1e004000 {
+		compatible = "mediatek,mt7621-crypto";
+		reg = <0x1e004000 0x1000>;
+
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_SHARED 19 IRQ_TYPE_LEVEL_HIGH>;
+	};
+
+	gic: interrupt-controller@1fbc0000 {
+		compatible = "mti,gic";
+		reg = <0x1fbc0000 0x80>;
+
+		interrupt-controller;
+		#interrupt-cells = <3>;
+
+		mti,reserved-cpu-vectors = <7>;
+
+		interrupt-parent = <&cpuintc>;
+
+		timer {
+			compatible = "mti,gic-timer";
+			interrupts = <GIC_LOCAL 1 IRQ_TYPE_NONE>;
+			clocks = <&cpuclock>;
+		};
+
+	};
+
+	ethernet@1e100000 {
+		compatible = "ralink,mt7621-eth";
+		reg = <0x1e100000 10000>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		resets = <&rstctrl 6 &rstctrl 23>;
+		reset-names = "fe", "eth";
+
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_SHARED 3 IRQ_TYPE_LEVEL_HIGH>;
+
+		mdio-bus {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			phy1f: ethernet-phy@1f {
+				reg = <0x1f>;
+				phy-mode = "rgmii";
+			};
+		};
+	};
+
+	gsw@1e110000 {
+		compatible = "ralink,mt7620a-gsw";
+		reg = <0x1e110000 8000>;
+	};
+
+	i2c@0 {
+	        compatible = "i2c-gpio";
+		sda-gpios = <&gpio0 3 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+		scl-gpios = <&gpio0 4 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+	};
+
+	pcie@1e140000 {
+		status = "disabled";
+		compatible = "mediatek,mt7621-pci";
+		reg = <0x1e140000 0x100
+			0x1e142000 0x100>;
+
+		#address-cells = <3>;
+		#size-cells = <2>;
+
+		device_type = "pci";
+
+		bus-range = <0 255>;
+		ranges = <
+			0x02000000 0 0x00000000 0x60000000 0 0x10000000 /* pci memory */
+			0x01000000 0 0x00000000 0x1e160000 0 0x00010000 /* io space */
+		>;
+
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_SHARED 4 IRQ_TYPE_LEVEL_HIGH
+				GIC_SHARED 24 IRQ_TYPE_LEVEL_HIGH
+				GIC_SHARED 25 IRQ_TYPE_LEVEL_HIGH>;
+
+		pcie@0,0 {
+			reg = <0x0000 0 0 0 0>;
+			#address-cells = <3>;
+			#size-cells = <2>;
+			ranges;
+			bus-range = <0x00 0xff>;
+		};
+
+		pcie@1,0 {
+			reg = <0x0800 0 0 0 0>;
+			#address-cells = <3>;
+			#size-cells = <2>;
+			ranges;
+			bus-range = <0x00 0xff>;
+		};
+
+		pcie@2,0 {
+			reg = <0x1000 0 0 0 0>;
+			#address-cells = <3>;
+			#size-cells = <2>;
+			ranges;
+			bus-range = <0x00 0xff>;
+		};
+	};
+};

```

</details>

<details>
<summary>ROS `prom.c` conditionally enables DTS nodes per board</summary>
</br>

```
+		} else if (mips_machtype == MACH_MT_LtAP) {
+			enable_device("/pcie@1e140000");
+			enable_device("/palmbus@1e000000/uartlite@e00");
+			enable_device("/palmbus@1e000000/spi@b00/m25p80@0");
+			enable_device("/palmbus@1e000000/spi@b00/ts@1");
+			enable_device("/sdhci@1e130000");
+			enable_device("/xhci@1e1c0000");
+			enable_device("/ltap-leds");
+			ports_prop.length = 1 * sizeof(unsigned);
+		}

```
</details>

If they don't work, I went digging for the 760 GPIOs by jailbreaking RouterOS, then watching which GPIOs changed in kernel when I changed software settings or hardware.

db4rne commented:
@john-tho thank you so much, this helped a lot. I didn't even think of looking for a GPL drop, now I know better.

I finished the general porting of the device now, only thing left is to look for switching the antenna  of the wifi chip. it has external connectors one could use, but for that we need to find the GPIOs connected to it first.

anyway, if someone could review this patch and tell me what to change it would be great.

f00b4r0 commented on target/linux/ramips/dts/mt7621_mikrotik_ltap-2hnd.dts:
```diff
@@ -0,0 +1,170 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led_pwr: pwr {
```
the `led_pwr:` alias is unneeded (it's not used)

f00b4r0 commented on target/linux/ramips/dts/mt7621_mikrotik_ltap-2hnd.dts:
```diff
@@ -0,0 +1,170 @@
+			gpios = <&gpio 0 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_gps: gps {
```
same with `led_gps:`

f00b4r0 reviewed (commented):
Two very minor cosmetic nits, otherwise LGTM.

One question about the commit message: doesn't the "common mikrotik procedure" (namely the reset button trick) apply? Usually summarized in commit as follows:

    Flashing:
     TFTP boot initramfs image and then perform sysupgrade. Follow common
     MikroTik procedure as in https://openwrt.org/toh/mikrotik/common.

db4rne commented:
@f00b4r0 thanks for your input. I updated the PR with your suggested changes and also changed the status LED to power, since that seems to be the LED more in line with OpenWRT.

f00b4r0 commented on target/linux/ramips/dts/mt7621_mikrotik_ltap-2hnd.dts:
```diff
@@ -0,0 +1,170 @@
+		led_pwr: pwr {
+			label = "blue:pwr";
+			gpios = <&gpio 23 GPIO_ACTIVE_HIGH>;
+			default-state = "keep";
```
The default-state should be "on" here.

f00b4r0 reviewed (changes requested):
Thanks for your update, however please revert the status LED change, you had it right the first time :)

db4rne commented:
I reverted the led changes and changed the status. again, thanks for you input

db4rne commented:
I rebased to latest master and incorporated some feedback by @lynxis

hauke commented on target/linux/ramips/dts/mt7621_mikrotik_ltap-2hnd.dts:
```diff
@@ -0,0 +1,171 @@
+		gpio = <&gpio 10 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-boot-on;
+		regulator-always-on;
```
Why do you haev to set this to `regulator-always-on;` ?

db4rne replied:
oh you are right, that was not needed. I removed it.

hauke commented on target/linux/ramips/dts/mt7621_mikrotik_ltap-2hnd.dts:
```diff
@@ -0,0 +1,169 @@
+#include "mt7621_mikrotik.dtsi"
```
Please add an SPDX license header like:
```
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```

hauke commented:
Please rebase this pull request to fix the github workflows.

db4rne commented:
I did a rebase and added the SPDX header

hauke commented:
Thank you for the patch, I applied it to master in ffa4b5283b390099a207dcd22c6e9b0b528ac1e4.

hnyman commented:
I think that this causes a bug, as it removed the case item closing ";;" from belkin,rt1800 and linksys,e7350 

https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/etc/board.d/02_network#L217-L225

``` 
	belkin,rt1800|\
	linksys,e7350)
		lan_mac=$(mtd_get_mac_ascii Config lan_hwaddr)
		wan_mac=$(mtd_get_mac_ascii Config wan_hwaddr)
		label_mac=$lan_mac
	mikrotik,ltap-2hnd)
		label_mac=$(cat "/sys/firmware/mikrotik/hard_config/mac_base")
		lan_mac=$label_mac
		;;
```

db4rne commented:
Yes, indeed that was lost, thanks for finding it. I fixed it in https://github.com/openwrt/openwrt/pull/11237 so maybe someone can merge that.
